engine/consensus/compliance/engine.go

```diff
@@ -399,3 +402,28 @@ func (e *Engine) BroadcastProposalWithDelay(header *flow.Header, delay time.Dura
 func (e *Engine) BroadcastProposal(header *flow.Header) error {
 	return e.BroadcastProposalWithDelay(header, 0)
 }
+
+// OnFinalizedBlock implements the `OnFinalizedBlock` callback from the `hotstuff.FinalizationConsumer`
+//  (1) Informs sealing.Core about finalization of respective block.
+// CAUTION: the input to this callback is treated as trusted; precautions should be taken that messages
+// from external nodes cannot be considered as inputs to this function
+func (e *Engine) OnFinalizedBlock(flow.Identifier) {
+	e.finalizationEventsNotifier.Notify()
+}
+
+// finalizationProcessingLoop is a separate goroutine that performs processing of finalization events
+func (e *Engine) finalizationProcessingLoop() {
+	finalizationNotifier := e.finalizationEventsNotifier.Channel()
+	for {
+		select {
+		case <-e.unit.Quit():
+			return
+		case <-finalizationNotifier:
+			finalized, err := e.state.Final().Head()
+			if err != nil {
+				e.log.Fatal().Err(err).Msg("could not retrieve last finalized block")
+			}
+			e.core.ProcessFinalizedBlock(finalized)
+		}
+	}
+}
```

Comment:
This approach works. 

But I have another idea here. The current approach queries the finalized block each time, we could technically get rid of this query if we consume [Protocol Event](https://github.com/onflow/flow-go/blob/master/state/protocol/events.go#L27), instead of [HotStuff Event](https://github.com/onflow/flow-go/blob/master/consensus/hotstuff/consumer.go#L31). The HotStuff Event only provides block view, but no block height, whereas the ProtocolEvent `OnBlockFinalized` provides the block header. 

But if we do so, we might have additional work to do: we don't want to block the consume call. 

Your current approach has done this by using the event notifier. Event Notifier will cache up to 1 unprocessed event, so that if it takes time to prune mempool/votecollectors, pushing a new event to the event notifier will not be blocked. 

If we want to get rid of the query by using the Protocol Event `OnBlockFinalized`, then we will need to implement a way that ensures:
1. the consumer should not block the protocol state `Mutate` call when processing `OnBlockFinalized` event.
2. If consumer is slow processing events, the event queue should keep the newest event which is the latest finalized block height, and drop the old event.

Thoughts?

Reply:
With current approach we satisfy both 1) and 2), it's the same approach we use in sealing engine. I believe it's not worth to over-engineer it at this moment since we are planning to update compliance engine either way.